kaolin/datasets/base.py

```diff
@@ -0,0 +1,154 @@
+            preprocessing_transform (Callable): Called on the outputs of _get_data over the indices
+                                                from 0 to len(self) during the construction of the dataset,
+                                                the preprocessed outputs are then cached to 'cache_dir'.
+            transform (Callable): Called on the preprocessed data at __getitem__.
+            no_progress (bool): disable tqdm progress bar for preprocessing.
+        """
+        self.initialize(*args, **kwargs)
+        if preprocessing_transform is not None:
+            desc = 'applying preprocessing'
+            if preprocessing_params is None:
+                preprocessing_params = {}
+            assert preprocessing_params.get('cache_dir') is not None
+            self.cache_convert = helpers.Cache(
+                preprocessing_transform, preprocessing_params['cache_dir'],
+                cache_key=helpers._get_hash(repr(preprocessing_transform)))
+            if preprocessing_params.get('use_cuda') is None:
+                preprocessing_params['use_cuda'] = False
+            num_workers = preprocessing_params.get('num_workers')
+            if num_workers == 0:
+                with torch.no_grad():
+                    for idx in tqdm(range(len(self)), desc=desc, disable=no_progress):
+                        name = self._get_attributes(idx)['name']
+                        if name not in self.cache_convert.cached_ids:
+                            data = self._get_data(idx)
+                            self.cache_convert(name, *data)
+            else:
+                p = Pool(num_workers)
+                iterator = p.imap_unordered(
+                    _preprocess_task,
+                    [(idx, self._get_data, self._get_attributes, self.cache_convert)
+                     for idx in range(len(self))])
+                for i in tqdm(range(len(self)), desc=desc, disable=no_progress):
+                    next(iterator)
+        else:
+            self.cache_convert = None
+        self.transform = transform
+
+    def __getitem__(self, index):
+        """Returns the item at index idx. """
+        attributes = self._get_attributes(index)
+        data = (self.cache_convert(attributes['name']) if self.cache_convert is not None else
+                self._get_data(index))
+        if self.transform is not None:
+            data = self.transform(data)
+        return {'data': data, 'attributes': attributes}
+
+    @abstractmethod
+    def initialize(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def _get_attributes(self, index):
+        pass
+
+    @abstractmethod
+    def _get_data(self, index):
+        pass
+
+    @abstractmethod
+    def __len__(self):
+        pass
+
+class CombinationDataset(KaolinDataset):
+    """Dataset combining multiple list of datasets outputs as inputs
```

Comment:
The description isn't super-clear imo

Reply:
Sure, can you suggest something ?

Reply:
It's tricky to explin, yeah. How about something like
```
Combines a list of datasets into a unified dataset object. Useful when multiple output representations are needed from a common base representation (Eg. when a mesh is to be served as both a pointcloud and a voxelgrid, etc.)
```